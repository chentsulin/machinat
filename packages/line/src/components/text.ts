/* eslint-disable import/prefer-default-export */
import { makePartSegment, PartSegment } from '@machinat/core/renderer';
import { annotateLineComponent } from '../utils';
import { LineComponent } from '../types';

/**
 * @category Props
 */
export type EmojiProps = {
  /**
   * Product ID for a set of LINE emoji. See Sendable
   * [LINE emoji list](https://d.line-scdn.net/r/devcenter/sendable_line_emoji_list.pdf).
   */
  productId: string;
  /**
   * ID for a LINE emoji inside a set. See Sendable
   * [LINE emoji list](https://d.line-scdn.net/r/devcenter/sendable_line_emoji_list.pdf).
   */
  emojiId: string;
};

const __Emoji = function Emoji(node, path) {
  const { productId, emojiId } = node.props;
  return [
    makePartSegment(node, path, {
      type: 'emoji_placeholder',
      productId,
      emojiId,
    }),
  ];
};

/**
 * Insert a LINE emoji within a {@link Expression} element.
 * @category Component
 * @props {@link EmojiProps}
 * @guides Check official [reference](https://developers.line.biz/en/reference/messaging-api/#text-message).
 */
export const Emoji: LineComponent<
  EmojiProps,
  PartSegment<any>
> = annotateLineComponent(__Emoji);
