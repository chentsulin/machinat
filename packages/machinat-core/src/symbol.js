// @flow
export const MACHINAT_ELEMENT_TYPE = Symbol('machinat.element');

export const MACHINAT_FRAGMENT_TYPE = Symbol('machinat.element.fragment');

export const MACHINAT_PAUSE_TYPE = Symbol('machinat.element.pause');

export const MACHINAT_PROVIDER_TYPE = Symbol('machinat.element.provider');

export const MACHINAT_THUNK_TYPE = Symbol('machinat.element.thunk');

export const MACHINAT_RAW_TYPE = Symbol('machinat.element.raw');

export const MACHINAT_NATIVE_TYPE = Symbol('machinat.component.native');

export const MACHINAT_SERVICES_CONTAINER = Symbol(
  'machinat.services.container'
);

export const MACHINAT_SERVICES_INTERFACE = Symbol(
  'machinat.services.interface'
);

export const MACHINAT_SERVICES_PROVIDER = Symbol('machinat.services.provider');
