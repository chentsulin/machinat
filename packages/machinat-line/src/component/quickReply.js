/* eslint-disable import/prefer-default-export */
import { valuesOfAssertedType } from 'machinat-utility';

import { asPartComponent } from './utils';

import * as _actionModule from './action';

const actionComponents = Object.values(_actionModule);
const getActionValues = valuesOfAssertedType(...actionComponents);

const QuickReply = ({ props: { imageURL, action } }, render) => {
  const actionValues = getActionValues(render(action, '.action'));

  return [
    {
      type: 'action',
      imageUrl: imageURL,
      action: actionValues && actionValues[0],
    },
  ];
};

const __QuickReply = asPartComponent(QuickReply);

export { __QuickReply as QuickReply };
