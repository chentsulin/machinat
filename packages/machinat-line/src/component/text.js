/* eslint-disable import/prefer-default-export */
import { textSegment } from 'machinat-renderer';
import { asContainerComponent } from '../utils';

const Emoji = async (node, _, path) => [
  textSegment(String.fromCodePoint(node.props.code), node, path),
];

const __Emoji = asContainerComponent(Emoji);

export { __Emoji as Emoji };
