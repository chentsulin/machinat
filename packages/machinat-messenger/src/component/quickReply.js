import { annotateNative } from './utils';

export const QuickReply = ({ title, payload, imageUrl }) => ({
  content_type: 'text',
  title,
  payload,
  image_url: imageUrl,
});
annotateNative(QuickReply);

const LOCATION_QUICK_REPLY = { content_type: 'location' };
export const LocationQuickReply = () => LOCATION_QUICK_REPLY;
annotateNative(LocationQuickReply);

const PHONE_QUICK_REPLY = { content_type: 'user_phone_number' };
export const PhoneQuickReply = () => PHONE_QUICK_REPLY;
annotateNative(PhoneQuickReply);

const EMAIL_QUICK_REPLY = { content_type: 'user_email' };
export const EmailQuickReply = () => EMAIL_QUICK_REPLY;
annotateNative(EmailQuickReply);
