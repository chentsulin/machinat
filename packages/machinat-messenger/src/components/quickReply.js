import { partSegment } from '@machinat/core/renderer';
import { annotateMessengerComponent } from '../utils';

export const QuickReply = (node, path) => {
  const { title, payload, imageURL } = node.props;
  return [
    partSegment(node, path, {
      content_type: 'text',
      title,
      payload,
      image_url: imageURL,
    }),
  ];
};
annotateMessengerComponent(QuickReply);

const PHONE_QUICK_REPLY_VALUES = { content_type: 'user_phone_number' };

export const PhoneQuickReply = (node, path) => [
  partSegment(node, path, PHONE_QUICK_REPLY_VALUES),
];
annotateMessengerComponent(PhoneQuickReply);

const EMAIL_QUICK_REPLY_VALUES = { content_type: 'user_email' };

export const EmailQuickReply = (node, path) => [
  partSegment(node, path, EMAIL_QUICK_REPLY_VALUES),
];
annotateMessengerComponent(EmailQuickReply);
