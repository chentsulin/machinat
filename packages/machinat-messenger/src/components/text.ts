/* eslint-disable import/prefer-default-export */
import type { MachinatNode } from '@machinat/core/types';
import formatNode from '@machinat/core/utils/formatNode';
import { textSegment } from '@machinat/core/renderer';
import type { TextSegment } from '@machinat/core/renderer/types';
import { annotateMessengerComponent } from '../utils';
import type { MessengerComponent } from '../types';

/**
 * @category Props
 */
type LatexProps = {
  /** Texual node to be wrapped in the LATEX block. */
  children: MachinatNode;
};

/** @ignore */
const LATEX_BEGIN = '\\(';

/** @ignore */
const LATEX_END = '\\)';

/** @ignore */
const _Latex = async function Latex(node, path, render) {
  const segments = await render(node.props.children, '.children');
  if (segments === null) {
    return null;
  }

  for (const segment of segments) {
    if (segment.type !== 'text') {
      throw new TypeError(
        `non-textual node ${formatNode(
          segment.node
        )} received, only textual nodes allowed`
      );
    }
  }

  return [
    textSegment(node, path, LATEX_BEGIN),
    segments[0],
    textSegment(node, path, LATEX_END),
  ];
};
/**
 * The receipt template allows you to send an order confirmation as a structured
 * message.
 * @category Component
 * @props {@link LatexProps}
 * @guides Check [help page](https://www.facebook.com/help/147348452522644).
 */
export const Latex: MessengerComponent<
  LatexProps,
  TextSegment
> = annotateMessengerComponent(_Latex);
