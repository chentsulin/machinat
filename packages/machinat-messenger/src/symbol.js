export const MESSENGER_NAITVE_TYPE = Symbol.for('machinat.messenger.native');
