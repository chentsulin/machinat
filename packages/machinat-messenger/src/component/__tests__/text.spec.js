import Machinat from '../../../../machinat';
import { MESSENGER_NAITVE_TYPE } from '../../symbol';
import { Text, Latex } from '../text';
import { QuickReply } from '../quickReply';
import renderHelper from './renderHelper';

const textNode = [<a>abcd</a>, <b>bcda</b>, <del>dabc</del>];
const quickReplies = [
  <QuickReply title="A" />,
  <QuickReply title="B" />,
  <QuickReply title="C" />,
];
const renderInside = jest.fn(
  node =>
    typeof node === 'string'
      ? [{ rendered: node, element: node }]
      : node === quickReplies
        ? [
            {
              rendered: '__RENDERED_QUICKREPLY_OBJ_1__',
              element: quickReplies[0],
            },
            {
              rendered: '__RENDERED_QUICKREPLY_OBJ_2__',
              element: quickReplies[1],
            },
            {
              rendered: '__RENDERED_QUICKREPLY_OBJ_3__',
              element: quickReplies[2],
            },
          ]
        : node === textNode
          ? [
              { rendered: '\n__RENDERED_TEXT_1__', element: textNode[0] },
              { rendered: '\n__RENDERED_TEXT_2__', element: textNode[1] },
              { rendered: '\n__RENDERED_TEXT_3__', element: textNode[2] },
            ]
          : undefined
);
const render = renderHelper(renderInside);

beforeEach(() => {
  renderInside.mockClear();
});

describe('Text', () => {
  it('is valid root Component', () => {
    expect(typeof Text).toBe('function');
    expect(Text.$$native).toBe(MESSENGER_NAITVE_TYPE);
    expect(Text.$$entry).toBe('me/messages');
    expect(Text.$$root).toBe(true);
  });

  it('match snapshot', () => {
    expect(
      [
        <Text>abc</Text>,
        <Text metadata="xyz">abc</Text>,
        <Text>{textNode}</Text>,
        <Text quickReplies={quickReplies}>{textNode}</Text>,
      ].map(render)
    ).toMatchSnapshot();
  });

  it('render children as "text" field', () => {
    expect(render(<Text>some text</Text>).message.text).toBe('some text');
    expect(render(<Text>{textNode}</Text>).message.text).toBe(`
__RENDERED_TEXT_1__
__RENDERED_TEXT_2__
__RENDERED_TEXT_3__`);
    expect(renderInside).toHaveBeenCalledWith('some text', '.children');
    expect(renderInside).toHaveBeenCalledWith(textNode, '.children');
  });

  it('render quickReplies prop as "quick_replies" field', () => {
    expect(
      render(<Text quickReplies={quickReplies}>abc</Text>).message.quick_replies
    ).toEqual([
      '__RENDERED_QUICKREPLY_OBJ_1__',
      '__RENDERED_QUICKREPLY_OBJ_2__',
      '__RENDERED_QUICKREPLY_OBJ_3__',
    ]);
    expect(renderInside).toHaveBeenCalledWith(quickReplies, '.quickReplies');
  });
});

describe('Latex', () => {
  it('is valid root Component', () => {
    expect(typeof Latex).toBe('function');
    expect(Latex.$$native).toBe(MESSENGER_NAITVE_TYPE);
    expect(Latex.$$entry).toBe(undefined);
    expect(Latex.$$root).toBe(undefined);
  });

  it('render children as "text" field', () => {
    expect(render(<Latex>some text</Latex>)).toBe('\\(some text\\)');
    expect(render(<Latex>{textNode}</Latex>)).toBe(`\\(
__RENDERED_TEXT_1__
__RENDERED_TEXT_2__
__RENDERED_TEXT_3__\\)`);
    expect(renderInside).toHaveBeenCalledWith('some text', '.children');
    expect(renderInside).toHaveBeenCalledWith(textNode, '.children');
  });
});
