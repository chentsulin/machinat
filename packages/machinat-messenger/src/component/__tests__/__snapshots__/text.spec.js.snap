// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text match snapshot 1`] = `
Array [
  Object {
    "message": Object {
      "metadata": undefined,
      "quick_replies": undefined,
      "text": "abc",
    },
  },
  Object {
    "message": Object {
      "metadata": "xyz",
      "quick_replies": undefined,
      "text": "abc",
    },
  },
  Object {
    "message": Object {
      "metadata": undefined,
      "quick_replies": undefined,
      "text": "
__RENDERED_TEXT_1__
__RENDERED_TEXT_2__
__RENDERED_TEXT_3__",
    },
  },
  Object {
    "message": Object {
      "metadata": undefined,
      "quick_replies": Array [
        "__RENDERED_QUICKREPLY_OBJ_1__",
        "__RENDERED_QUICKREPLY_OBJ_2__",
        "__RENDERED_QUICKREPLY_OBJ_3__",
      ],
      "text": "
__RENDERED_TEXT_1__
__RENDERED_TEXT_2__
__RENDERED_TEXT_3__",
    },
  },
]
`;
